chronos/armory_refunds.py

```diff
@@ -1,86 +1,414 @@
-        next(csvread, None)  # Header
-        for row in csvread:
-            if row[1]:
-                locker.append([int(row[0]), row[1].lower(), int(row[2])])  # uid, classname, count
+    # get playerid to uid mapping
+    with connect(host=args.databasehost, database=DB_USERS, user=args.databaseuser, password=args.databasepassword) as conn:
+        cursor = conn.cursor()
+
+        cursor.execute(f"SELECT entity_id, field_player_id_value FROM {TABLE_IDS}")
+        for (uid, playerid) in cursor:
+            players[playerid] = uid
+            players[uid] = playerid
+
+    # find removed items and assemble all data
+    with connect(host=args.databasehost, database=DB_NAME, user=args.databaseuser, password=args.databasepassword) as conn:
+        cursor = conn.cursor()
+
+        cursor.execute(f"SELECT className, prettyName, price FROM {TABLE_ITEMLIST}")
+        for (classname, prettyname, price) in cursor:
+            classname = classname.lower()
+            item_list[classname] = Item(classname, prettyname, int(price))
+
+        # locker
+        cursor.execute(f"SELECT uid, className, quantity FROM {TABLE_LOCKER}")
+        for (uid, classname, quantity) in cursor:
+            classname = classname.lower()
+            if classname not in item_list and quantity > 0:
+                refunds[uid].uid = uid
+                if classname in old_item_list:
+                    refunds[uid].total += quantity * old_item_list[classname].price
+                    refunds[uid].items.append(old_item_list[classname])
+                    refunds[uid].note.append(f"{quantity}x {old_item_list[classname].prettyname}")
+                else:
+                    refunds[uid].items.append(Item(classname, "", 0))
+                    print(f"warning: '{classname}' owned by {uid} not found in old item list - it will be removed without refund")
+                    #input("  confirm?")
+
+        # loadout
+        cursor.execute(f"SELECT playerID, loadout FROM {TABLE_EQUIPPED}")
+        for (playerid, loadout_orig) in cursor:
+            uid = players[playerid]
+            refunds[uid].uid = uid
+
+            loadout_orig = loadout_orig.lower()
+            loadout = loadout_from_arma(loadout_orig)
+
+            for old_item in old_item_list.values():
+                if old_item.classname not in item_list.keys():
+                    while (fnr := loadout.find_and_remove(old_item.classname))[0] > 0:
+                        quantity, removed = fnr
+                        if not removed:
+                            print(f"warning: unable to remove '{old_item.classname}' from {uid} ({playerid}) - item will be refunded, but not removed!")
+                            found_index = loadout_orig.find(old_item.classname)
+                            print(f"    {loadout_orig[max(0, found_index - 15):found_index + len(old_item.classname) + 100]}")
+                            input("  confirm?")
```

Comment:
What if the user wants to cancel at this point. "confirm?" makes one believe he can deny this (yet he can't ?)

Reply:
He can `Ctrl+C`, since nothing is being applied at this point yet.

Reply:
Personally, I would add that in parenthesis or something. You never know who's going to run this.